Assert that mutexes and prints are not use in interrupt or critical context

c1728p9 opened the pull request:
Add a function to determine if you are in ISR context, add  asserts for context to mutex and printf usage and increase error buffer size so long asserts are printed correctly.

c1728p9 commented:
This requires #4388 to pass testing.

theotherjimmy commented:
Are the changes to the RTX header files going to conflict with https://github.com/ARMmbed/mbed-os/pull/4294 ?

c1728p9 commented:
@theotherjimmy I'll hold off a bit on merging this to prevent conflicts with #4294. Worst case, if I need to merge this I'll rebase #4294 myself.

0xc0170 commented:
retest uvisor

bulislaw commented:
RTX2 is doing the checks (if isr) by itself eg https://github.com/ARMmbed/mbed-os/blob/feature_cmsis5/rtos/rtx5/TARGET_CORTEX_M/rtx_mutex.c#L457

0xc0170 commented:
> RTX2 is doing the checks (if isr) by itself eg https://github.com/ARMmbed/mbed-os/blob/feature_cmsis5/rtos/rtx5/TARGET_CORTEX_M/rtx_mutex.c#L457

I forgot that they added events so users should be find out what is with the system. Based on this, should assert be only in C++, and keep RTX as it is (my above comment) ?

bulislaw commented:
I think there's no need to modify RTX5.

pan- commented:
I'm not sure to understand the purpose of this PR. The functions `osMutexWait` and `osMutexRelease` already return the error `osErrorISR` if they are called from interrupt context. 

Adding assertion at this point do **break** the API. Not running in ISR it not a precondition for calling these function (even if it will fail), with this change, it become one. Users referring to CMSIS RTOS documentation might fall into that trap where a valid code become invalid because of that change. 

For low level `_read` and `_write` I wonder how much user code it will break and what is the impact on mbed OS 2 application.

c1728p9 commented:
Hi @pan-, the goal here is to trap on errors immediately if the code is wrong. Using a mutex in an interrupt handler indicates that there is an error in the program. In mbed-os 2 calling _read and _write in an interrupt handler resulted in undefined behavior.

If you have any good use cases in mind where it is valid to use a mutex in an interrupt handler let me know, as that means this PR would be a problematic breaking change to the API.

sg- commented:
Looks like this needs a conflict resolved and to only be enabled when in debug build. This change should not modify behavior of default or release builds.

c1728p9 commented:
Rebased on top of RTX5, changed ```core_util_in_isr``` to ```core_util_is_isr_active```, enabled trapping all supported RTX errors, and added the define MBED_TRAP_ERRORS_ENABLED to enable the trapping behavior. By default error trapping is only enabled in debug builds.

Only top 9 commits are relevent. The rest are are for RTX5. This PR needs to be rebased on top of the main line RTX5 changes before running CI.

sg- commented:
/morph test
